src/replication/stream_senders_manager.cpp

```diff
@@ -0,0 +1,244 @@
+  const unsigned int stream_senders_manager::SUPERVISOR_DAEMON_CHECK_CONN_MS = 5000;
+
+  stream_senders_manager::stream_senders_manager (cubstream::stream &supervised_stream):
+    m_stream (supervised_stream)
+  {
+    init ();
+  }
+
+  stream_senders_manager::~stream_senders_manager ()
+  {
+    finalize ();
+  }
+
+  void stream_senders_manager::init ()
+  {
+#if defined (SERVER_MODE)
+    int error_code = NO_ERROR;
+
+    stream_senders.clear ();
+
+    std::string daemon_name = "senders_supervisor_daemon_" + m_stream.name ();
+    auto exec_f = std::bind (&stream_senders_manager::execute, this, std::placeholders::_1);
+    cubthread::entry_callable_task *task = new cubthread::entry_callable_task (exec_f);
+    supervisor_daemon = cubthread::get_manager ()->create_daemon (
+				cubthread::looper (std::chrono::milliseconds (SUPERVISOR_DAEMON_DELAY_MS)), task, daemon_name.c_str ());
+
+    error_code = rwlock_initialize (&senders_lock, "MASTER_SENDERS_LOCK");
+    assert (error_code == NO_ERROR);
+#endif
+  }
+
+  void stream_senders_manager::add_stream_sender (cubstream::transfer_sender *sender)
+  {
+    rwlock_write_lock (&senders_lock);
+    stream_senders.push_back (sender);
+    rwlock_write_unlock (&senders_lock);
+  }
+
+  void stream_senders_manager::stop_stream_sender (cubstream::transfer_sender *sender)
+  {
+    rwlock_write_lock (&senders_lock);
+    for (cubstream::transfer_sender *s : stream_senders)
+      {
+	if (sender == s)
+	  {
+	    er_log_debug_replication (ARG_FILE_LINE, "stream_senders_manager::stop_stream_sender for channel:%s",
+				      s->get_channel ().get_channel_id ().c_str ());
+	    s->enter_termination_phase ();
+	    break;
+	  }
+      }
+    rwlock_write_unlock (&senders_lock);
+  }
+
+
+  bool stream_senders_manager::find_stream_sender (const cubstream::transfer_sender *sender)
+  {
+    std::vector<cubstream::transfer_sender *>::iterator it;
+    bool found = false;
+
+    rwlock_read_lock (&senders_lock);
+    for (cubstream::transfer_sender *s : stream_senders)
+      {
+	if (sender == s)
+	  {
+	    found = true;
+	    break;
+	  }
+      }
+    rwlock_read_unlock (&senders_lock);
+
+    return found;
+  }
+
+  void stream_senders_manager::finalize ()
+  {
+#if defined (SERVER_MODE)
+    er_log_debug_replication (ARG_FILE_LINE, "stream_senders_manager::finalize");
+
+    int error_code = NO_ERROR;
+
+    if (supervisor_daemon != NULL)
+      {
+	cubthread::get_manager ()->destroy_daemon (supervisor_daemon);
+	supervisor_daemon = NULL;
+      }
+
+    rwlock_write_lock (&senders_lock);
+    for (cubstream::transfer_sender *sender : stream_senders)
+      {
+	delete sender;
+      }
+    stream_senders.clear ();
+    rwlock_write_unlock (&senders_lock);
+
+    error_code = rwlock_finalize (&senders_lock);
+    assert (error_code == NO_ERROR);
+#endif
+  }
+
+  std::size_t stream_senders_manager::get_number_of_stream_senders ()
```

Comment:
const member function

Reply:
it calls rwlock_read_lock (which changes a data member)